src/mapreduce.jl

```diff
@@ -0,0 +1,276 @@
+end
+
+function tname(s::Symbol)
+    Symbol("t", s)
+end
+
+const SYMBOLS_MAPREDUCE_LIKE = [:sum, :prod, :minimum, :maximum]
+for fun in SYMBOLS_MAPREDUCE_LIKE
+    tfun = tname(fun)
+    docstring = _make_docstring(fun, ["[f]", "src::AbstractArray"])
+    @eval begin
+        export $tfun
+        @doc $docstring ->
+        function $tfun end
+    end
+end
+
+@doc _make_docstring(:map!, ["f", "dst::AbstractArray", "srcs::AbstractArray..."]) ->
+function tmap! end
+
+@doc _make_docstring(:map, ["f", "srcs::AbstractArray..."]) ->
+function tmap end
+
+@doc _make_docstring(:mapreduce, ["f", "op", "src::AbstractArray"], ["[init]"]) ->
+function tmapreduce end
+
+@doc _make_docstring(:reduce, ["op", "src::AbstractArray"],  ["[init]"]) ->
+function treduce end
+
+############################## Helper functions ##############################
+struct Batches{V}
+    firstindex::Int
+    lastindex::Int
+    batch_size::Int
+    values::V
+    length::Int
+end
+
+function Batches(values, batch_size::Integer)
+    @argcheck batch_size > 0
+    r = eachindex(values)
+    @assert r isa AbstractUnitRange
+    len = ceil(Int, length(r) / batch_size)
+    Batches(first(r), last(r), batch_size, values, len)
+end
+
+Base.length(o::Batches) = o.length
+Base.eachindex(o::Batches) = Base.OneTo(length(o))
+function Base.getindex(o::Batches, i)
+    @boundscheck (@argcheck i in eachindex(o))
+    start = o.firstindex + (i-1) * o.batch_size
+    stop  = min(start + (o.batch_size) -1, o.lastindex)
+    o.values[start:stop]
+end
+
+function default_batch_size(len)
+    len <= 1 && return 1
+    nthreads=Threads.nthreads()
+    items_per_thread = len/nthreads
+    items_per_batch = items_per_thread/4
+    clamp(1, round(Int, items_per_batch), len)
+end
+
+function Base.iterate(o::Batches, state=1)
+    if state in eachindex(o)
+        o[state], state+1
+    else
+        nothing
+    end
+end
+
+mutable struct _RollingCutOut{A,I<:AbstractUnitRange,T} <: AbstractVector{T}
+    array::A
+    eachindex::I
+end
+
+function _RollingCutOut(array::AbstractArray, indices)
+    T = eltype(array)
+    A = typeof(array)
+    I = typeof(indices)
+    _RollingCutOut{A, I, T}(array, indices)
+end
+
+Base.size(r::_RollingCutOut) = (length(r.eachindex),)
+
+function Base.eachindex(r::_RollingCutOut, rs::_RollingCutOut...)
+    for r2 in rs
+        @assert r.eachindex == r2.eachindex
+    end
+    Base.IdentityUnitRange(r.eachindex)
+end
+Base.axes(r::_RollingCutOut) = (eachindex(r),)
+
+@inline function Base.getindex(o::_RollingCutOut, i)
+    @boundscheck checkbounds(o, i)
+    @inbounds o.array[i]
+end
+
+@inline function Base.setindex!(o::_RollingCutOut, val, i)
+    @boundscheck checkbounds(o, i)
+    @inbounds o.array[i] = val
+end
+
+############################## tmap, tmap! ##############################
+struct MapWorkspace{F,B,AD,AS}
+    f::F
+    batches::B
+    arena_dst_view::AD
+    arena_src_views::AS
+end
+
+@noinline function run!(o::MapWorkspace)
+    let o=o
```

Comment:
Why not use the current callable struct approach which is guaranteed to infer instead of the magical `let`?

Reply:
You mean I create an ad hoc wrapper struct that is called in the loop instead?

Reply:
No, I mean to just make `MapWorkspace` itself callable but that's just a matter of preference. Perhaps call it `TMap` or something. `(::TMap)(args...)` will then have this function body. This is also a pattern that facilitates advanced users to pre-allocate the workspace and reuse it without having to reach in for the `run!` function.

Reply:
Note that the argument of a `TMap` object would then be a batch index. I think that is confusing and advanced users would just have to dig as deep into internals as calling `run!`.
I could add a comment about the `let` trick instead ok?

Reply:
I meant for the whole loop to be inside the call to `(::TMap)(args...)`. This will look cleaner imo, compared to the `let` workaround. Adding a comment is also ok, but please check that `run!` actually infers properly since sometimes the `let` trick doesn't work with `@threads`.

Reply:
Ok I don't quite understand what you mean. Can you maybe post a code snippet of the `(::TMap)(args...)` function? And explain what is `args...` here?

Reply:
Oh in that case `args...` are not needed. The function can look like this:
```julia
function (o::TMap)()
	Threads.@threads for i in 1:length(o.batches)
	    tid = Threads.threadid()
	    dst_view  = o.arena_dst_view[tid]
	    src_views = o.arena_src_views[tid]
	    inds = o.batches[i]
	    dst_view.eachindex = inds
	    for view in src_views
	        view.eachindex = inds
	    end
	    Base.map!(o.f, dst_view, src_views...)
	end
end
```